### DISCUSSION TITLE
Recommendation towards usage as a global data object

### DISCUSSION THREAD

**Discussion by maxacarvalho** [spatie/laravel-data#340] [Q&A]
Hi,
First of all, thank you for such amazing package!

I'm planning to use the package as a central data object solution, here's an example.

## Let's think about a blog post data object.
```php
class PostData extends Data
{
    public function __construct(
        #[Rule('required')]
        public int $id,
        #[Rule('required', 'uuid')]
        public string $uuid,
        #[Rule('required', 'string')]
        public string $title,
        #[Rule('required', 'string')]
        public string $content
    ) {
    }
}
```

That data object will be able to replace:
* A DataTransferObject
* A FormRequest
* An API resource

But then, I have some doubts.

### When creating a post

Validation would fail since I don't have an `ID` nor a `UUID` yet.
```php
class StorePostController extends Controller
{
    public function __invoke(
        PostData $postData,
        CreatePostAction $action
    ) {
        try {
            $action->handle($postData);
            
            return response()->json([], Response::HTTP_CREATED);
        } catch (Throwable $exception) {
            return response()->json([], Response::HTTP_INTERNAL_SERVER_ERROR);
        }
    }
}
```

### When listing posts
To solve the situation above I could remove the `ID` and `UUID` from the data object but that would cause the listing to fails since I want at least one of the identifiers.

```php
class IndexPostController extends Controller
{
    public function __invoke() {
        return PostData::collection(Post::all());
    }
}
```

----

I wonder if you have any similar use case that you can share?

Thanks!

**Comment by ahoiroman** [comment]
You could either create a DataObject for the create action - or, I did not try this myself yet, make `id` lazy with the condition of the existence of entry.

Any ideas on that?
